src/main/java/com/gradle/enterprise/conventions/customvalueprovider/GradleEnterpriseConventions.java

```diff
@@ -0,0 +1,145 @@
+
+
+import com.gradle.scan.plugin.BuildScanExtension;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+
+import javax.annotation.Nullable;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.gradle.enterprise.conventions.customvalueprovider.ScanCustomValueNames.GIT_COMMIT_NAME;
+
+public class GradleEnterpriseConventions {
+    private static final Logger LOGGER = Logging.getLogger(GradleEnterpriseConventions.class);
+    private static final String PUBLIC_GRADLE_ENTERPRISE_SERVER = "https://ge.gradle.org";
+    private static final String GRADLE_ENTERPRISE_URL_PROPERTY_NAME = "gradle.enterprise.url";
+    private static final String CI_ENV_NAME = "CI";
+
+    private static final Pattern HTTPS_URL_PATTERN = Pattern.compile("https://github\\.com/([\\w-]+)/([\\w-]+)\\.git");
+    private static final Pattern SSH_URL_PATTERN = Pattern.compile("git@github\\.com:([\\w-]+)/([\\w-]+)\\.git");
+
+    private final ProviderFactory providerFactory;
+    private final String gradleEnterpriseServerUrl;
+    private final boolean isCiServer;
+
+    public GradleEnterpriseConventions(ProviderFactory providerFactory) {
+        this.providerFactory = providerFactory;
+        this.gradleEnterpriseServerUrl = getSystemProperty(GRADLE_ENTERPRISE_URL_PROPERTY_NAME, PUBLIC_GRADLE_ENTERPRISE_SERVER, providerFactory);
+        this.isCiServer = !providerFactory.environmentVariable(CI_ENV_NAME).forUseAtConfigurationTime().getOrElse("").isEmpty();
+
+    }
+
+    public String customValueSearchUrl(Map<String, String> search) {
+        String query = search.entrySet()
+            .stream()
+            .map(entry -> String.format("search.names=%s&search.values=%s", urlEncode(entry.getKey()), urlEncode(entry.getValue())))
+            .collect(Collectors.joining("&"));
+        return String.format("%s/scans?%s", gradleEnterpriseServerUrl, query);
+    }
+
+    public String getGradleEnterpriseServerUrl() {
+        return gradleEnterpriseServerUrl;
+    }
+
+    public boolean isCiServer() {
+        return isCiServer;
+    }
+
+    public String getSystemProperty(String name, String defaultValue) {
+        return getSystemProperty(name, defaultValue, getProviderFactory());
+    }
+
+    private static String getSystemProperty(String name, String defaultValue, ProviderFactory providerFactory) {
+        return providerFactory.systemProperty(name).forUseAtConfigurationTime().orElse(defaultValue).get();
```

Comment:
```suggestion
        return systemPropertyProvider(name).orElse(defaultValue).get();
```

Reply:
That doesn't work since this method is static so it can be used from the constructor. `systemPropertyProvider` isn't static.